src/future/shared.rs

```diff
@@ -105,188 +126,165 @@ impl<F> Future for Shared<F>
     type Error = SharedError<F::Error>;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let mut state = self.inner.state.lock().unwrap();
-        let unparker = match *state {
-            State::Waiting(ref unparker, ref _original_future) => {
-                let mut unparker_inner = unparker.inner.lock().unwrap();
-                if unparker_inner.original_future_needs_poll {
-                    unparker_inner.original_future_needs_poll = false;
-                    unparker.clone()
-                } else {
-                    unparker_inner.insert(self.id, task::park());
-                    return Ok(Async::NotReady)
-                }
+        self.set_waiter();
```

Comment:
It is not needed in the complete branch, but avoiding it adds somewhat significant code. I am favoring simpler code here in this PR and improvements can be done in later PRs.